### DISCUSSION TITLE
App Tries to Create Log File in Root Private Folder

### DISCUSSION THREAD

**Discussion by tk27182** [pyinstaller/pyinstaller#7296] [Help] [answered]
I've built an app bundle for MacOS using Pyinstaller that I'm beta-testing with colleagues. After building on my local machine, I'm able to open and run the app with no issues. However, when my colleagues download the release and try to run the app, the app is unable to connect to the server. When I check the system.log file, I see the error

```
Failed to execute script 'run_waitress' due to unhandled exception: [Errno 30] Read-only file system: 
'/private/var/folders/vf/3xw3wws956q2sqm5zx7w2n800000gp/T/AppTranslocation/  
B746854D-70C1-4B5B-8A27-1F2547B0637C/d/error.log'
APP[5366]: Traceback:
Traceback (most recent call last):
	File "run_waitress.py", line 46, in <module>
	File "logging/__init__.py", line 1984, in basicConfig
	File "logging/__init__.py", line 1143, in __init__
	File "logging/__init__.py", line 1172, in _open
OSError: [Errno 30] Read-only file system: '/private/var/folders/vf/3xw3wws956q2sqm5zx7w2n800000gp/  
T/AppTranslocation/B746854D-70C1-4B5B-8A27-1F2547B0637C/d/error.log'
```

I'm serving the app using `waitress` and I wanted to create a logging file to get more details when issues arise. The logging file is created in my `run_waitress.py` script that the bundle is built around, and is configured like

```
logger = logging.getLogger('waitress')
logging.basicConfig(filename='error.log',
                                  filemode='w',
                                  level=logging.INFO
                                 )
```

I checked my Systems Preferences first to confirm it's set to allow apps from the Store and the web (it is).

I've tried specifying the path in different ways to try to not get it to write there but to no avail. When I run the app locally, it creates the `error.log` file in the folder where the app lives, as desired. How can I tell the app not to write the log in the root private folder, but in the folder where the app lives?

**Comment by rokm** [comment, marked as answer]
MacOS app bundles are supposed to be immutable, so you cannot write into their directories. To enforce this at runtime, the OS translocates the bundle to a read-only destination - in this specific case case `/private/var/folders/vf/3xw3wws956q2sqm5zx7w2n800000gp/T/AppTranslocation/B746854D-70C1-4B5B-8A27-1F2547B0637C/d`.

> filename='error.log'

You'll need to change this to use an absolute path, probably somewhere in the user's home directory. You can use [appdirs](https://pypi.org/project/appdirs/) to determine what the suitable directory would be.

**Reply by tk27182** [reply]
Thank you, that works! 

But what prevents me from defining the absolute path to the log file like this? It seems like it should work.

```
import sys
import re
from pathlib import Path

if getattr(sys, 'frozen', False):
    # we are running in a bundle
    application_path = Path(sys._MEIPASS)
else:
    # we are running in a normal Python environment
    application_path = Path(__file__)

m = re.search('.+?(?=APPNAME\.)', str(application_path))
if m is None:
    # normal Python environment
    new_path = application_path.parent
else:
    new_path = m.group(0)

log_path = Path(new_path).joinpath('error.log')
```

**Reply by rokm** [reply]
> But what prevents me from defining the absolute path to the log file like this? It seems like it should work.

The OS, which translocates (a copy of) your app bundle to a read-only location *prior to running it* (try printing or displaying `sys._MEIPASS` or `__file__` or `sys.executable` to see that the bundle is not anymore where you think it is...). So all attempts at writing into application bundle (e.g., next to the executable) are going to be met with `Read-only file system` errors.

**Reply by tk27182** [reply]
Thanks, I see that!
